### PR TITLE
`istio-cni` - stop writing `kubeconfig` to shared CNI netdir

### PR DESCRIPTION
**Pull request by bleggett**
In https://github.com/istio/istio/issues/52273 I realized that CNI_NET_DIR as an env var is

- not referenced directly anywhere at runtime
- not used by the CNI agent at all as a runtime env
- is used in one spot at runtime - the CNI plugin. Which uses it to find the kubeconfig it uses to build a kube client - which means we have a config sitting in the netdir, that has the netdir as a path, so the thing invoked via the config can read its own config and get the netdir path...etc etc etc. 

None of that is really necessary, and it isn't great for several reasons:

- It's hard to explain and understand
- It's hard to test
- CNI_NET_DIR is platform-dependent path variable because it's part of the CNI spec, but the `kubeconfig` we write out isn't - we're the only ones that use it.
- Writing `kubeconfig` to CNI_NET_DIR isn't really necessary, and tends to trigger a lot of extra watch churns for us and for any other daemonset that might be watching that commonly-watched spot.

So, this puts all our hostmount `sock` files as well as the kubeconfig in the same path in `/var/run/istio-cni` (where we were already putting both sock files by default)

This path is configurable, and we will put all that stuff in the provided path, but it doesn't need to be tied up with the shared CNI config location, and we can hardcode the actual filenames we want for stuff that only we populate + use.
 
This also removes several config flags which the user would effectively never need to override, and which we mostly just ignore in tests.

This means the `cni net dir` is set purely in Helm, and is used to establish a well-known pod mount for the agent to dump files into - and nothing else.

From a security perspective:

putting the `kubeconfig` next to the sockets is fine - the credentials are used only by the client of those sockets (our plugin) to talk to the thing that dumped its own svcAccount token into that config in the first place. If anything it's a little better since we aren't advertising our agent's pod credentials to every CNI watcher on the node - but this is probably moot, everything we're talking about (CNI plugins, container runtime, CNI agents) are going to be at least as privileged as we are if not more so by definition.
 
 tl;dr
- Don't require passing `CNI_NET_DIR` config to the plugin just so it can load a non-shared `kubeconfig` from that (shared) location which has nothing to do with this flow.
- Put all the host node files we require 
  - log sock
  - cni event sock
  - kubecreds
in the same spot (except for CNI config + bins which has a standard spot)
- cleanup config:
  - Add: `CNIAgentRunDir: /var/run/istio-cni` user config val.
  - Remove: `LogUDSAddress` user config val
  - Remove: `CNIEventAddress` user config val
  - Remove: `CNINetDir` user config val
  - Remove: `Kubernetes.kubeconfig` user config val (only exists as a unit test punchthru anyway)

https://github.com/istio/istio/issues/52315

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
LGTM

cc @jwendell

**Comment by jwendell** [comment]
I'll try this on openshift tomorrow

**Comment by jwendell** [comment]
This doesn't work on openshift.

Installed with
```
$ ./out/linux_amd64/istioctl install --set profile=openshift --set hub=gcr.io/istio-testing --set tag=latest
```

Control plane deploys fine.

Tried to deploy the sleep with automatic sidecar injection, pod remains at Init status:
```
$ oc get pod
NAME                     READY   STATUS     RESTARTS   AGE
sleep-7656cf8794-vr8bt   0/2     Init:0/1   0          89s
```

`oc describe pod` last line:
```
 Warning  FailedCreatePodSandBox  7s                kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_sleep-7656cf8794-vr8bt_sleep_b5630ec6-fb6b-450d-b44a-3d0694eadd4e_0(b2a5185945eaa95402917a2e51bdd47dabd551b1b8617d2cea02a2c771bbb2e0): error adding pod sleep_sleep-7656cf8794-vr8bt to CNI network "multus-cni-network": plugin type="multus-shim" name="multus-cni-network" failed (add): CmdAdd (shim): CNI request failed with status 400: 'ContainerID:"b2a5185945eaa95402917a2e51bdd47dabd551b1b8617d2cea02a2c771bbb2e0" Netns:"/var/run/netns/2547f9f0-1217-44fc-a4bb-9a6057635832" IfName:"eth0" Args:"IgnoreUnknown=1;K8S_POD_NAMESPACE=sleep;K8S_POD_NAME=sleep-7656cf8794-vr8bt;K8S_POD_INFRA_CONTAINER_ID=b2a5185945eaa95402917a2e51bdd47dabd551b1b8617d2cea02a2c771bbb2e0;K8S_POD_UID=b5630ec6-fb6b-450d-b44a-3d0694eadd4e" Path:"" ERRORED: error configuring pod [sleep/sleep-7656cf8794-vr8bt] networking: [sleep/sleep-7656cf8794-vr8bt/b5630ec6-fb6b-450d-b44a-3d0694eadd4e:istio-cni]: error adding container to network "istio-cni": the server could not find the requested resource (get pods sleep-7656cf8794-vr8bt)
': StdinData: {"binDir":"/var/lib/cni/bin","clusterNetwork":"/host/run/multus/cni/net.d/10-ovn-kubernetes.conf","cniVersion":"0.3.1","daemonSocketDir":"/run/multus/socket","globalNamespaces":"default,openshift-multus,openshift-sriov-network-operator","logLevel":"verbose","logToStderr":true,"name":"multus-cni-network","namespaceIsolation":true,"type":"multus-shim"}
```

**Comment by bleggett** [comment]
@jwendell can you get me the `istio-cni` log? Something is odd here.

**Comment by bleggett** [comment]
I also _cannot_ reproduce this failure with `k3d` with sidecars or ambient - which overrides the CNI bin and conf dirs like `openshift` does, so I kinda wanna dig into what's going on here.

**Comment by jwendell** [comment]
[logb.log](https://github.com/user-attachments/files/16415751/logb.log)

**Comment by jwendell** [comment]
If you are willing to try it yourself, you can download `crc`,it's a light openshift that runs locally, and uses virtualization: https://console.redhat.com/openshift/create/local

**Comment by bleggett** [comment]
> [logb.log](https://github.com/user-attachments/files/16415751/logb.log)

> wrote kubeconfig file /host/etc/cni/net.d/ZZZ-istio-cni-kubeconfig

Hm, this implies either the version you were using isn't built from this branch. In this branch we don't write the kubeconfig to `/host/etc/cni/net.d/` anymore and we aren't writing to `ZZZ-istio-cni-kubeconfig` - I dropped the `ZZZ` prefix, it's just hardcoded to `istio-cni-kubeconfig` now.

**Comment by bleggett** [comment]
Lemme test myself tho to be 100% sure.

**Comment by jwendell** [comment]
That's right, I was using hub=gcr.io/istio-testing and tag=latest. Let me try again.

**Comment by jwendell** [comment]
You are right. My bad. I should have tested with the built images, not with images on gcr.

Now it's working. Sorry.

**Comment by bleggett** [comment]
> You are right. My bad. I should have tested with the built images, not with images on gcr.
> 
> Now it's working. Sorry.

Awesome. Thanks for confirming, and testing! 

And sorry for the churn on this, as I ever-so-slowly actually figured out what the heck was happening and why.

**Review by jwendell — APPROVED** [review]

File: cni/pkg/install/kubeconfig.go [hunk 1]
```diff
@@ -124,40 +125,19 @@ func createKubeConfig(cfg *config.InstallConfig) (kubeconfig, error) {
...
-// or if a kubeconfig exists there, but differs from the current config.
-// In any case, an error indicates the file must be (re)written, and no error means no action need be taken
-func checkExistingKubeConfigFile(cfg *config.InstallConfig, expectedKC kubeconfig) error {
-	kubeconfigFilepath := filepath.Join(cfg.MountedCNINetDir, cfg.KubeconfigFilename)
-
-	existingKC, err := os.ReadFile(kubeconfigFilepath)
-	if err != nil {
-		installLog.Debugf("no preexisting kubeconfig at %s, assuming we need to create one", kubeconfigFilepath)
+	kubeconfigFilepath := filepath.Join(cfg.CNIAgentRunDir, constants.CNIPluginKubeconfName)
+	if err := file.AtomicWrite(kubeconfigFilepath, []byte(kc.Full), os.FileMode(cfg.KubeconfigMode)); err != nil {
```
Comment (MorrisLaw) [inline cni/pkg/install/kubeconfig.go:138]: should we consider logging the error with debugf?

**Review by howardjohn — APPROVED** [review]

File: releasenotes/notes/52323.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 52315
+
+releaseNotes:
+- |
+  **Removed** Stop writing kubeconfig to CNI net dir. Remove `CNI_NET_DIR` from the `istio-cni` configmap, as it now does nothing.
```
Comment (dhawton) [inline releasenotes/notes/52323.yaml:9]:
```suggestion
  **Removed** writing kubeconfig to CNI net directory.
- |
  **Removed** `CNI_NET_DIR` from the `istio-cni` configmap, as it now does nothing.
```
